Java Storage NIO:  FileSystemProvider::checkAccess fails on '/' with StorageException

Using com.google.cloud:google-cloud-nio:0.124.20 the CloudStorageFileSystemProvider implementation of FileSystemProvider.checkAccess fails when asking about the root directory.

Take this code fragment and fill in a cloud storage bucket name you have access to:

```java
public class App {

    private static final Logger LOG = LoggerFactory.getLogger(App.class);

    public static void main(String[] args) {

        String bucketName = "xxxxxxxxxxxx";

        FileSystem fileSystem = FileSystems.getFileSystem(URI.create("gs://" + bucketName));

        Path path = fileSystem.getPath("/");

        FileSystemProvider provider = fileSystem.provider();
        try {
            provider.checkAccess(path, AccessMode.READ, AccessMode.WRITE);
        }
        catch (Exception e) {
            LOG.error("{}", e);
        }
        LOG.warn("{}", path);
    }
}
```

#### Stack trace
NOTE: I have hidden the real bucket name.

```
com.google.cloud.storage.StorageException: Invalid field selection bucket
	at com.google.cloud.storage.StorageException.translate(StorageException.java:163)
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.translate(HttpStorageRpc.java:291)
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.get(HttpStorageRpc.java:517)
	at com.google.cloud.storage.StorageImpl.lambda$get$6(StorageImpl.java:298)
	at com.google.api.gax.retrying.DirectRetryingExecutor.submit(DirectRetryingExecutor.java:103)
	at com.google.cloud.RetryHelper.run(RetryHelper.java:76)
	at com.google.cloud.RetryHelper.runWithRetries(RetryHelper.java:50)
	at com.google.cloud.storage.Retrying.run(Retrying.java:60)
	at com.google.cloud.storage.StorageImpl.run(StorageImpl.java:1476)
	at com.google.cloud.storage.StorageImpl.get(StorageImpl.java:296)
	at com.google.cloud.storage.contrib.nio.CloudStorageFileSystemProvider.checkAccess(CloudStorageFileSystemProvider.java:732)
	at nl.basjes.bugreports.App.main(App.java:31)
Caused by: com.google.api.client.googleapis.json.GoogleJsonResponseException: 400 Bad Request
GET https://storage.googleapis.com/storage/v1/b/xxxxxxxxxxxx/o/?fields=bucket,name,id&projection=full
{
  "code" : 400,
  "errors" : [ {
    "domain" : "global",
    "location" : "fields",
    "locationType" : "parameter",
    "message" : "Invalid field selection bucket",
    "reason" : "invalidParameter"
  } ],
  "message" : "Invalid field selection bucket"
}
	at com.google.api.client.googleapis.json.GoogleJsonResponseException.from(GoogleJsonResponseException.java:146)
	at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:118)
	at com.google.api.client.googleapis.services.json.AbstractGoogleJsonClientRequest.newExceptionOnError(AbstractGoogleJsonClientRequest.java:37)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest$1.interceptResponse(AbstractGoogleClientRequest.java:439)
	at com.google.api.client.http.HttpRequest.execute(HttpRequest.java:1111)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:525)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.executeUnparsed(AbstractGoogleClientRequest.java:466)
	at com.google.api.client.googleapis.services.AbstractGoogleClientRequest.execute(AbstractGoogleClientRequest.java:576)
	at com.google.cloud.storage.spi.v1.HttpStorageRpc.get(HttpStorageRpc.java:514)
	... 9 more
```

#### External references such as API reference guides

- https://docs.oracle.com/javase/8/docs/api/java/nio/file/spi/FileSystemProvider.html#checkAccess-java.nio.file.Path-java.nio.file.AccessMode...-

#### Any additional information below

I ran into this problem when "simply" trying to attach Apache Mina SFTP to a Google Cloud Bucket.
Note that this all does work as expected on 
- existing files/directories
- non-existing files/directories

Only on '/' it fails.

--- a/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemProvider.java
+++ b/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemProvider.java
@@ -34,6 +34,7 @@ import com.google.cloud.storage.StorageException;
 import com.google.cloud.storage.StorageOptions;
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.base.MoreObjects;
+import com.google.common.base.Optional;
 import com.google.common.collect.AbstractIterator;
 import com.google.common.net.UrlEscapers;
 import com.google.common.primitives.Ints;
@@ -62,15 +63,23 @@ import java.nio.file.attribute.BasicFileAttributeView;
 import java.nio.file.attribute.BasicFileAttributes;
 import java.nio.file.attribute.FileAttribute;
 import java.nio.file.attribute.FileAttributeView;
+import java.nio.file.attribute.FileTime;
+import java.nio.file.attribute.GroupPrincipal;
+import java.nio.file.attribute.PosixFileAttributes;
+import java.nio.file.attribute.PosixFilePermission;
+import java.nio.file.attribute.UserPrincipal;
 import java.nio.file.spi.FileSystemProvider;
 import java.util.ArrayList;
+import java.util.Arrays;
 import java.util.Collections;
+import java.util.EnumSet;
 import java.util.HashMap;
 import java.util.Iterator;
 import java.util.List;
 import java.util.Map;
 import java.util.Objects;
 import java.util.Set;
+import java.util.TreeMap;
 import javax.annotation.Nullable;
 import javax.annotation.concurrent.ThreadSafe;
 import javax.inject.Singleton;
@@ -839,11 +848,91 @@ public final class CloudStorageFileSystemProvider extends FileSystemProvider {
   }
 
   @Override
-  public Map<String, Object> readAttributes(Path path, String attributes, LinkOption... options) {
+  public Map<String, Object> readAttributes(Path path, String attributes, LinkOption... options) throws IOException {
     // TODO(#811): Java 7 NIO defines at least eleven string attributes we'd want to support
     //             (eg. BasicFileAttributeView and PosixFileAttributeView), so rather than a partial
     //             implementation we rely on the other overload for now.
-    throw new UnsupportedOperationException();
+
+    // Partial implementation for a few commonly used ones: basic, gcs
+    String[] split = attributes.split(":", 2);
+    if (split.length != 2) {
+      throw new UnsupportedOperationException();
+    }
+    String view = split[0];
+    List<String> attributeNames = Arrays.asList(split[1].split(","));
+    boolean allAttributes = attributeNames.size() == 1 && attributeNames.get(0).equals("*");
+
+    BasicFileAttributes fileAttributes;
+
+    Map<String, Object> results = new TreeMap<>();
+    switch (view) {
+      case "gcs":
+        fileAttributes = readAttributes(path, CloudStorageFileAttributes.class, options);
+        break;
+      case "basic":
+        fileAttributes = readAttributes(path, BasicFileAttributes.class, options);
+        break;
+      default:
+        throw new UnsupportedOperationException();
+    }
+
+    if (fileAttributes == null) {
+      throw new UnsupportedOperationException();
+    }
+
+    // BasicFileAttributes
+    if (allAttributes || attributeNames.contains("lastModifiedTime")) {
+      results.put("lastModifiedTime", fileAttributes.lastModifiedTime());
+    }
+    if (allAttributes || attributeNames.contains("lastAccessTime")) {
+      results.put("lastAccessTime", fileAttributes.lastAccessTime());
+    }
+    if (allAttributes || attributeNames.contains("creationTime")) {
+      results.put("creationTime", fileAttributes.creationTime());
+    }
+    if (allAttributes || attributeNames.contains("isRegularFile")) {
+      results.put("isRegularFile", fileAttributes.isRegularFile());
+    }
+    if (allAttributes || attributeNames.contains("isDirectory")) {
+      results.put("isDirectory", fileAttributes.isDirectory());
+    }
+    if (allAttributes || attributeNames.contains("isSymbolicLink")) {
+      results.put("isSymbolicLink", fileAttributes.isSymbolicLink());
+    }
+    if (allAttributes || attributeNames.contains("isOther")) {
+      results.put("isOther", fileAttributes.isOther());
+    }
+    if (allAttributes || attributeNames.contains("size")) {
+      results.put("size", fileAttributes.size());
+    }
+
+    // CloudStorageFileAttributes
+    if (fileAttributes instanceof CloudStorageFileAttributes) {
+      CloudStorageFileAttributes cloudStorageFileAttributes = (CloudStorageFileAttributes) fileAttributes;
+      if (allAttributes || attributeNames.contains("etag")) {
+        results.put("etag", cloudStorageFileAttributes.etag());
+      }
+      if (allAttributes || attributeNames.contains("mimeType")) {
+        results.put("mimeType", cloudStorageFileAttributes.mimeType());
+      }
+      if (allAttributes || attributeNames.contains("acl")) {
+        results.put("acl", cloudStorageFileAttributes.acl());
+      }
+      if (allAttributes || attributeNames.contains("cacheControl")) {
+        results.put("cacheControl", cloudStorageFileAttributes.cacheControl());
+      }
+      if (allAttributes || attributeNames.contains("contentEncoding")) {
+        results.put("contentEncoding", cloudStorageFileAttributes.contentEncoding());
+      }
+      if (allAttributes || attributeNames.contains("contentDisposition")) {
+        results.put("contentDisposition", cloudStorageFileAttributes.contentDisposition());
+      }
+      if (allAttributes || attributeNames.contains("userMetadata")) {
+        results.put("userMetadata", cloudStorageFileAttributes.userMetadata());
+      }
+    }
+
+    return results;
   }
 
   @Override

--- a/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemProvider.java
+++ b/google-cloud-nio/src/main/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemProvider.java
@@ -34,7 +34,6 @@ import com.google.cloud.storage.StorageException;
 import com.google.cloud.storage.StorageOptions;
 import com.google.common.annotations.VisibleForTesting;
 import com.google.common.base.MoreObjects;
-import com.google.common.base.Optional;
 import com.google.common.collect.AbstractIterator;
 import com.google.common.net.UrlEscapers;
 import com.google.common.primitives.Ints;
@@ -63,16 +62,10 @@ import java.nio.file.attribute.BasicFileAttributeView;
 import java.nio.file.attribute.BasicFileAttributes;
 import java.nio.file.attribute.FileAttribute;
 import java.nio.file.attribute.FileAttributeView;
-import java.nio.file.attribute.FileTime;
-import java.nio.file.attribute.GroupPrincipal;
-import java.nio.file.attribute.PosixFileAttributes;
-import java.nio.file.attribute.PosixFilePermission;
-import java.nio.file.attribute.UserPrincipal;
 import java.nio.file.spi.FileSystemProvider;
 import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Collections;
-import java.util.EnumSet;
 import java.util.HashMap;
 import java.util.Iterator;
 import java.util.List;
@@ -848,7 +841,8 @@ public final class CloudStorageFileSystemProvider extends FileSystemProvider {
   }
 
   @Override
-  public Map<String, Object> readAttributes(Path path, String attributes, LinkOption... options) throws IOException {
+  public Map<String, Object> readAttributes(Path path, String attributes, LinkOption... options)
+      throws IOException {
     // TODO(#811): Java 7 NIO defines at least eleven string attributes we'd want to support
     //             (eg. BasicFileAttributeView and PosixFileAttributeView), so rather than a partial
     //             implementation we rely on the other overload for now.
@@ -908,7 +902,8 @@ public final class CloudStorageFileSystemProvider extends FileSystemProvider {
 
     // CloudStorageFileAttributes
     if (fileAttributes instanceof CloudStorageFileAttributes) {
-      CloudStorageFileAttributes cloudStorageFileAttributes = (CloudStorageFileAttributes) fileAttributes;
+      CloudStorageFileAttributes cloudStorageFileAttributes =
+          (CloudStorageFileAttributes) fileAttributes;
       if (allAttributes || attributeNames.contains("etag")) {
         results.put("etag", cloudStorageFileAttributes.etag());
       }

--- a/google-cloud-nio/src/test/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemProviderTest.java
+++ b/google-cloud-nio/src/test/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemProviderTest.java
@@ -64,7 +64,6 @@ import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
 import java.util.TreeMap;
-
 import org.junit.After;
 import org.junit.Assert;
 import org.junit.Before;
@@ -836,81 +835,87 @@ public class CloudStorageFileSystemProviderTest {
 
     BasicFileAttributes attributesBasic = mock(BasicFileAttributes.class);
     // BasicFileAttributes
-    when(attributesBasic.creationTime())        .thenReturn(FileTime.fromMillis(1L));
-    when(attributesBasic.lastModifiedTime())    .thenReturn(FileTime.fromMillis(2L));
-    when(attributesBasic.lastAccessTime())      .thenReturn(FileTime.fromMillis(3L));
-    when(attributesBasic.isRegularFile())       .thenReturn(true);
-    when(attributesBasic.isDirectory())         .thenReturn(true);
-    when(attributesBasic.isSymbolicLink())      .thenReturn(true);
-    when(attributesBasic.isOther())             .thenReturn(true);
-    when(attributesBasic.size())                .thenReturn(42L);
+    when(attributesBasic.creationTime()).thenReturn(FileTime.fromMillis(1L));
+    when(attributesBasic.lastModifiedTime()).thenReturn(FileTime.fromMillis(2L));
+    when(attributesBasic.lastAccessTime()).thenReturn(FileTime.fromMillis(3L));
+    when(attributesBasic.isRegularFile()).thenReturn(true);
+    when(attributesBasic.isDirectory()).thenReturn(true);
+    when(attributesBasic.isSymbolicLink()).thenReturn(true);
+    when(attributesBasic.isOther()).thenReturn(true);
+    when(attributesBasic.size()).thenReturn(42L);
 
     CloudStorageFileAttributes attributesGcs = mock(CloudStorageFileAttributes.class);
     // BasicFileAttributes
-    when(attributesGcs.creationTime())        .thenReturn(FileTime.fromMillis(1L));
-    when(attributesGcs.lastModifiedTime())    .thenReturn(FileTime.fromMillis(2L));
-    when(attributesGcs.lastAccessTime())      .thenReturn(FileTime.fromMillis(3L));
-    when(attributesGcs.isRegularFile())       .thenReturn(true);
-    when(attributesGcs.isDirectory())         .thenReturn(true);
-    when(attributesGcs.isSymbolicLink())      .thenReturn(true);
-    when(attributesGcs.isOther())             .thenReturn(true);
-    when(attributesGcs.size())                .thenReturn(42L);
+    when(attributesGcs.creationTime()).thenReturn(FileTime.fromMillis(1L));
+    when(attributesGcs.lastModifiedTime()).thenReturn(FileTime.fromMillis(2L));
+    when(attributesGcs.lastAccessTime()).thenReturn(FileTime.fromMillis(3L));
+    when(attributesGcs.isRegularFile()).thenReturn(true);
+    when(attributesGcs.isDirectory()).thenReturn(true);
+    when(attributesGcs.isSymbolicLink()).thenReturn(true);
+    when(attributesGcs.isOther()).thenReturn(true);
+    when(attributesGcs.size()).thenReturn(42L);
 
     List<Acl> acls = ImmutableList.of(Acl.newBuilder(new User("Foo"), OWNER).build());
 
     // CloudStorageFileAttributes
-    when(attributesGcs.etag())                .thenReturn(Optional.of("TheEtag"));
-    when(attributesGcs.mimeType())            .thenReturn(Optional.of("TheMimeType"));
-    when(attributesGcs.acl())                 .thenReturn(Optional.of(acls));
-    when(attributesGcs.cacheControl())        .thenReturn(Optional.of("TheCacheControl"));
-    when(attributesGcs.contentEncoding())     .thenReturn(Optional.of("TheContentEncoding"));
-    when(attributesGcs.contentDisposition())  .thenReturn(Optional.of("TheContentDisposition"));
-    when(attributesGcs.userMetadata())        .thenReturn(new TreeMap<>());
+    when(attributesGcs.etag()).thenReturn(Optional.of("TheEtag"));
+    when(attributesGcs.mimeType()).thenReturn(Optional.of("TheMimeType"));
+    when(attributesGcs.acl()).thenReturn(Optional.of(acls));
+    when(attributesGcs.cacheControl()).thenReturn(Optional.of("TheCacheControl"));
+    when(attributesGcs.contentEncoding()).thenReturn(Optional.of("TheContentEncoding"));
+    when(attributesGcs.contentDisposition()).thenReturn(Optional.of("TheContentDisposition"));
+    when(attributesGcs.userMetadata()).thenReturn(new TreeMap<>());
 
     CloudStoragePath path1 = CloudStoragePath.getPath(fileSystem, "/");
-    when(fileSystemProvider.readAttributes(path1, BasicFileAttributes.class))       .thenReturn(attributesBasic);
-    when(fileSystemProvider.readAttributes(path1, CloudStorageFileAttributes.class)).thenReturn(attributesGcs);
+    when(fileSystemProvider.readAttributes(path1, BasicFileAttributes.class))
+        .thenReturn(attributesBasic);
+    when(fileSystemProvider.readAttributes(path1, CloudStorageFileAttributes.class))
+        .thenReturn(attributesGcs);
 
     Map<String, Object> expectedBasic = new TreeMap<>();
     // BasicFileAttributes
-    expectedBasic.put("creationTime",       FileTime.fromMillis(1L));
-    expectedBasic.put("lastModifiedTime",   FileTime.fromMillis(2L));
-    expectedBasic.put("lastAccessTime",     FileTime.fromMillis(3L));
-    expectedBasic.put("isRegularFile",      true);
-    expectedBasic.put("isDirectory",        true);
-    expectedBasic.put("isSymbolicLink",     true);
-    expectedBasic.put("isOther",            true);
-    expectedBasic.put("size",               42L);
+    expectedBasic.put("creationTime", FileTime.fromMillis(1L));
+    expectedBasic.put("lastModifiedTime", FileTime.fromMillis(2L));
+    expectedBasic.put("lastAccessTime", FileTime.fromMillis(3L));
+    expectedBasic.put("isRegularFile", true);
+    expectedBasic.put("isDirectory", true);
+    expectedBasic.put("isSymbolicLink", true);
+    expectedBasic.put("isOther", true);
+    expectedBasic.put("size", 42L);
 
     assertEquals(expectedBasic, fileSystemProvider.readAttributes(path1, "basic:*"));
 
     Map<String, Object> expectedGcs = new TreeMap<>(expectedBasic);
     // CloudStorageFileAttributes
-    expectedGcs.put("etag",                 Optional.of("TheEtag"));
-    expectedGcs.put("mimeType",             Optional.of("TheMimeType"));
-    expectedGcs.put("acl",                  Optional.of(acls));
-    expectedGcs.put("cacheControl",         Optional.of("TheCacheControl"));
-    expectedGcs.put("contentEncoding",      Optional.of("TheContentEncoding"));
-    expectedGcs.put("contentDisposition",   Optional.of("TheContentDisposition"));
-    expectedGcs.put("userMetadata",         new TreeMap<>());
+    expectedGcs.put("etag", Optional.of("TheEtag"));
+    expectedGcs.put("mimeType", Optional.of("TheMimeType"));
+    expectedGcs.put("acl", Optional.of(acls));
+    expectedGcs.put("cacheControl", Optional.of("TheCacheControl"));
+    expectedGcs.put("contentEncoding", Optional.of("TheContentEncoding"));
+    expectedGcs.put("contentDisposition", Optional.of("TheContentDisposition"));
+    expectedGcs.put("userMetadata", new TreeMap<>());
 
     assertEquals(expectedGcs, fileSystemProvider.readAttributes(path1, "gcs:*"));
 
     Map<String, Object> expectedSpecific = new TreeMap<>();
-    expectedSpecific.put("lastModifiedTime",      FileTime.fromMillis(2L));
-    expectedSpecific.put("isSymbolicLink",        true);
-    expectedSpecific.put("isOther",               true);
+    expectedSpecific.put("lastModifiedTime", FileTime.fromMillis(2L));
+    expectedSpecific.put("isSymbolicLink", true);
+    expectedSpecific.put("isOther", true);
 
     // Asking for attributes that should NOT be known because we ask for basic view !
-    assertEquals(expectedSpecific, fileSystemProvider.readAttributes(path1,
-            "basic:lastModifiedTime,isSymbolicLink,isOther,etag,cacheControl"));
+    assertEquals(
+        expectedSpecific,
+        fileSystemProvider.readAttributes(
+            path1, "basic:lastModifiedTime,isSymbolicLink,isOther,etag,cacheControl"));
 
     // Add the attributes that are only known in gcs view
-    expectedSpecific.put("etag",                  Optional.of("TheEtag"));
-    expectedSpecific.put("cacheControl",          Optional.of("TheCacheControl"));
+    expectedSpecific.put("etag", Optional.of("TheEtag"));
+    expectedSpecific.put("cacheControl", Optional.of("TheCacheControl"));
 
-    assertEquals(expectedSpecific, fileSystemProvider.readAttributes(path1,
-            "gcs:lastModifiedTime,isSymbolicLink,isOther,etag,cacheControl"));
+    assertEquals(
+        expectedSpecific,
+        fileSystemProvider.readAttributes(
+            path1, "gcs:lastModifiedTime,isSymbolicLink,isOther,etag,cacheControl"));
   }
 
   private static CloudStorageConfiguration permitEmptyPathComponents(boolean value) {

--- a/google-cloud-nio/src/test/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemProviderTest.java
+++ b/google-cloud-nio/src/test/java/com/google/cloud/storage/contrib/nio/CloudStorageFileSystemProviderTest.java
@@ -16,6 +16,7 @@
 
 package com.google.cloud.storage.contrib.nio;
 
+import static com.google.cloud.storage.Acl.Role.OWNER;
 import static com.google.cloud.storage.contrib.nio.CloudStorageFileSystem.forBucket;
 import static com.google.common.truth.Truth.assertThat;
 import static java.nio.charset.StandardCharsets.UTF_8;
@@ -26,10 +27,17 @@ import static java.nio.file.StandardOpenOption.CREATE;
 import static java.nio.file.StandardOpenOption.CREATE_NEW;
 import static java.nio.file.StandardOpenOption.TRUNCATE_EXISTING;
 import static java.nio.file.StandardOpenOption.WRITE;
+import static org.junit.Assert.assertEquals;
 import static org.junit.Assert.assertTrue;
+import static org.mockito.Mockito.mock;
+import static org.mockito.Mockito.spy;
+import static org.mockito.Mockito.when;
 
+import com.google.cloud.storage.Acl;
+import com.google.cloud.storage.Acl.User;
 import com.google.cloud.storage.contrib.nio.testing.LocalStorageHelper;
 import com.google.cloud.testing.junit4.MultipleAttemptsRule;
+import com.google.common.base.Optional;
 import com.google.common.collect.ImmutableList;
 import com.google.common.testing.NullPointerTester;
 import java.io.IOException;
@@ -49,10 +57,14 @@ import java.nio.file.NoSuchFileException;
 import java.nio.file.OpenOption;
 import java.nio.file.Path;
 import java.nio.file.Paths;
+import java.nio.file.attribute.BasicFileAttributes;
+import java.nio.file.attribute.FileTime;
 import java.util.ArrayList;
 import java.util.HashMap;
 import java.util.List;
 import java.util.Map;
+import java.util.TreeMap;
+
 import org.junit.After;
 import org.junit.Assert;
 import org.junit.Before;
@@ -815,6 +827,90 @@ public class CloudStorageFileSystemProviderTest {
   public void getUserAgentStartsWithCorrectToken() {
     assertThat(String.format("gcloud-java-nio/%s", StorageOptionsUtil.USER_AGENT_ENTRY_VERSION))
         .startsWith("gcloud-java-nio/");
+  }
+
+  @Test
+  public void testReadAttributes() throws IOException {
+    CloudStorageFileSystem fileSystem = forBucket("dummy");
+    CloudStorageFileSystemProvider fileSystemProvider = spy(fileSystem.provider());
+
+    BasicFileAttributes attributesBasic = mock(BasicFileAttributes.class);
+    // BasicFileAttributes
+    when(attributesBasic.creationTime())        .thenReturn(FileTime.fromMillis(1L));
+    when(attributesBasic.lastModifiedTime())    .thenReturn(FileTime.fromMillis(2L));
+    when(attributesBasic.lastAccessTime())      .thenReturn(FileTime.fromMillis(3L));
+    when(attributesBasic.isRegularFile())       .thenReturn(true);
+    when(attributesBasic.isDirectory())         .thenReturn(true);
+    when(attributesBasic.isSymbolicLink())      .thenReturn(true);
+    when(attributesBasic.isOther())             .thenReturn(true);
+    when(attributesBasic.size())                .thenReturn(42L);
+
+    CloudStorageFileAttributes attributesGcs = mock(CloudStorageFileAttributes.class);
+    // BasicFileAttributes
+    when(attributesGcs.creationTime())        .thenReturn(FileTime.fromMillis(1L));
+    when(attributesGcs.lastModifiedTime())    .thenReturn(FileTime.fromMillis(2L));
+    when(attributesGcs.lastAccessTime())      .thenReturn(FileTime.fromMillis(3L));
+    when(attributesGcs.isRegularFile())       .thenReturn(true);
+    when(attributesGcs.isDirectory())         .thenReturn(true);
+    when(attributesGcs.isSymbolicLink())      .thenReturn(true);
+    when(attributesGcs.isOther())             .thenReturn(true);
+    when(attributesGcs.size())                .thenReturn(42L);
+
+    List<Acl> acls = ImmutableList.of(Acl.newBuilder(new User("Foo"), OWNER).build());
+
+    // CloudStorageFileAttributes
+    when(attributesGcs.etag())                .thenReturn(Optional.of("TheEtag"));
+    when(attributesGcs.mimeType())            .thenReturn(Optional.of("TheMimeType"));
+    when(attributesGcs.acl())                 .thenReturn(Optional.of(acls));
+    when(attributesGcs.cacheControl())        .thenReturn(Optional.of("TheCacheControl"));
+    when(attributesGcs.contentEncoding())     .thenReturn(Optional.of("TheContentEncoding"));
+    when(attributesGcs.contentDisposition())  .thenReturn(Optional.of("TheContentDisposition"));
+    when(attributesGcs.userMetadata())        .thenReturn(new TreeMap<>());
+
+    CloudStoragePath path1 = CloudStoragePath.getPath(fileSystem, "/");
+    when(fileSystemProvider.readAttributes(path1, BasicFileAttributes.class))       .thenReturn(attributesBasic);
+    when(fileSystemProvider.readAttributes(path1, CloudStorageFileAttributes.class)).thenReturn(attributesGcs);
+
+    Map<String, Object> expectedBasic = new TreeMap<>();
+    // BasicFileAttributes
+    expectedBasic.put("creationTime",       FileTime.fromMillis(1L));
+    expectedBasic.put("lastModifiedTime",   FileTime.fromMillis(2L));
+    expectedBasic.put("lastAccessTime",     FileTime.fromMillis(3L));
+    expectedBasic.put("isRegularFile",      true);
+    expectedBasic.put("isDirectory",        true);
+    expectedBasic.put("isSymbolicLink",     true);
+    expectedBasic.put("isOther",            true);
+    expectedBasic.put("size",               42L);
+
+    assertEquals(expectedBasic, fileSystemProvider.readAttributes(path1, "basic:*"));
+
+    Map<String, Object> expectedGcs = new TreeMap<>(expectedBasic);
+    // CloudStorageFileAttributes
+    expectedGcs.put("etag",                 Optional.of("TheEtag"));
+    expectedGcs.put("mimeType",             Optional.of("TheMimeType"));
+    expectedGcs.put("acl",                  Optional.of(acls));
+    expectedGcs.put("cacheControl",         Optional.of("TheCacheControl"));
+    expectedGcs.put("contentEncoding",      Optional.of("TheContentEncoding"));
+    expectedGcs.put("contentDisposition",   Optional.of("TheContentDisposition"));
+    expectedGcs.put("userMetadata",         new TreeMap<>());
+
+    assertEquals(expectedGcs, fileSystemProvider.readAttributes(path1, "gcs:*"));
+
+    Map<String, Object> expectedSpecific = new TreeMap<>();
+    expectedSpecific.put("lastModifiedTime",      FileTime.fromMillis(2L));
+    expectedSpecific.put("isSymbolicLink",        true);
+    expectedSpecific.put("isOther",               true);
+
+    // Asking for attributes that should NOT be known because we ask for basic view !
+    assertEquals(expectedSpecific, fileSystemProvider.readAttributes(path1,
+            "basic:lastModifiedTime,isSymbolicLink,isOther,etag,cacheControl"));
+
+    // Add the attributes that are only known in gcs view
+    expectedSpecific.put("etag",                  Optional.of("TheEtag"));
+    expectedSpecific.put("cacheControl",          Optional.of("TheCacheControl"));
+
+    assertEquals(expectedSpecific, fileSystemProvider.readAttributes(path1,
+            "gcs:lastModifiedTime,isSymbolicLink,isOther,etag,cacheControl"));
   }
 
   private static CloudStorageConfiguration permitEmptyPathComponents(boolean value) {

